django/db/models/lookups.py

```diff
@@ -297,6 +297,19 @@ def as_postgresql(self, compiler, connection):
         params = tuple(lhs_params) + tuple(rhs_params)
         return '%s %s %s' % (lhs, self.postgres_operator, rhs), params
 
+    @classmethod
+    def from_constraint(cls, lhs, rhs, operator):
+        from django.contrib.postgres.indexes import OpClass
+
+        # We need to get rid of OpClass because it only has sense during the constraint creation
+        args = [
+            expr.get_source_expressions()[0] if isinstance(expr, OpClass) else expr
+            for expr in (lhs, rhs)
+        ]
+        instance = cls(*args)
+        instance.postgres_operator = operator
+        return instance
+
```

Comment:
This should live on `PostgresOperatorLookup` and not `Lookup` itself; both `OpClass` and `ExclusionConstraint` are `django.contrib.postgres` construct.

Reply:
NVM, just realized this is actually `PostgresOperatorLookup` 🤦 I was under the impression that it lived in `django.contrib.postgres.lookups` but it seems to be relied on by `JSONField`.

Maybe that it would be better to confine the logic in `ExclusionConstraint.validate` then 🤷

Reply:
Shouldn't `PostgresOperatorLookup` also be in `django.contrib.postgres` then?

Reply:
Possibly but this patch is already touching a lot of areas so I'd suggest moving this whole `from_constraint` logic back to `ExclusionConstraint.validate` instead.

Sorry for the back and forth here.

Reply:
> Shouldn't `PostgresOperatorLookup` also be in `django.contrib.postgres` then?

No, because then `django.db` will depend on `django.contrib`.

Also, it looks like JSON support is being baked into the next version of SQLite, and it'll support some PostgreSQL-style operators too. So `PostgresOperatorLookup` could become more generic.